File: api/v1/schemas/picker_backtest.py
# -*- coding: utf-8 -*-
"""Picker backtest API schemas."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class PickerBacktestRunRequest(BaseModel):
    start_date: str = Field(..., description="Start date YYYY-MM-DD or YYYYMMDD")
    end_date: str = Field(..., description="End date YYYY-MM-DD or YYYYMMDD")
    hold_days: int = Field(10, ge=1, le=60, description="Holding period in trading days")
    top_n: int = Field(5, ge=1, le=20, description="Number of picks per day by score")
    picker_strategies: Optional[List[str]] = Field(
        None,
        description="Strategies to use: buy_pullback, breakout, bottom_reversal, eod_buyback",
    )


class PickerBacktestResultItem(BaseModel):
    trade_date: str
    code: str
    name: str = ""
    entry_price: float
    exit_price: Optional[float] = None
    return_pct: Optional[float] = None
    outcome: str  # win | loss | insufficient
    score: float = 0.0


class PickerBacktestSummary(BaseModel):
    start_date: str
    end_date: str
    hold_days: int
    top_n: int
    trade_dates_with_picks: int = 0  # days that had candidates (explains low total_picks)
    total_picks: int
    win_count: int
    loss_count: int
    insufficient_count: int
    win_rate_pct: Optional[float] = None
    avg_return_pct: Optional[float] = None
    max_drawdown_pct: Optional[float] = None
    profit_factor: Optional[float] = None
    alpha_vs_benchmark_pct: Optional[float] = None
    benchmark_avg_return_pct: Optional[float] = None


class PickerBacktestRunResponse(BaseModel):
    success: bool = True
    results: List[PickerBacktestResultItem] = Field(default_factory=list)
    summary: Optional[PickerBacktestSummary] = None
    trade_dates_count: int = 0

File: apps/dsa-web/src/types/pickerBacktest.ts
/**
 * Picker backtest API types
 * Mirrors api/v1/schemas/picker_backtest.py
 */

export type PickerStrategy = 'buy_pullback' | 'breakout' | 'bottom_reversal' | 'reversal_breakout' | 'small_cap' | 'slow_bull';

export interface PickerBacktestRunRequest {
  startDate: string;
  endDate: string;
  holdDays?: number;
  topN?: number;
  pickerStrategies?: PickerStrategy[];
}

export interface PickerBacktestResultItem {
  tradeDate: string;
  code: string;
  name?: string;
  entryPrice: number;
  exitPrice?: number;
  returnPct?: number;
  outcome: string;
  score?: number;
  exitReason?: string;  // stop_loss / trailing_ma10 / stage_break_+12pct / hardcap_+20pct / window_end / ...
  holdDays?: number;
  strategyId?: string;  // buy_pullback / breakout / bottom_reversal / reversal_breakout / small_cap / slow_bull
}

export interface PickerBacktestSummary {
  startDate: string;
  endDate: string;
  holdDays: number;
  topN: number;
  tradeDatesWithPicks?: number;  // days that had candidates
  totalPicks: number;
  winCount: number;
  lossCount: number;
  insufficientCount: number;
  winRatePct?: number;
  avgReturnPct?: number;
  maxDrawdownPct?: number;
  profitFactor?: number;
  alphaVsBenchmarkPct?: number;
  benchmarkAvgReturnPct?: number;
  cagrPct?: number;
  sharpeRatio?: number;
  calmarRatio?: number;
}

export interface PickerBacktestRunResponse {
  success: boolean;
  results: PickerBacktestResultItem[];
  summary: PickerBacktestSummary | null;
  tradeDatesCount: number;
}

export interface PickerBacktestHistoryItem {
  id: number;
  startDate: string;
  endDate: string;
  holdDays: number;
  topN: number;
  pickerStrategies?: string[];
  tradeDatesCount: number;
  winRatePct?: number;
  avgReturnPct?: number;
  createdAt?: string;
}
